### PR TITLE
chore: added type information to `core/field_dropdown.ts`

### PR DESCRIPTION
**Pull request by btw17**
## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)
- [x] I ran `npm run format` and `npm run lint`

### Proposed Changes

Provided type information in place of `AnyDuringMigration` for `core/field_dropdown.ts`

#### Behavior Before Change

Less type information.

#### Behavior After Change

More type information.

### Reason for Changes

Cleaning up types that weren't included in the TS migration.

### Test Coverage

N/A

### Documentation

Yes - just the automatically generated documentation that utilizes the TS types.

### Additional Information

I added comments to the file changes with anything of note. There are a couple of spots where the implementation of a few functions has been updated, though it shouldn't impact the functionality.

### CODE REVIEW

File: core/field_dropdown.ts
Comment (btw17) [inline core/field_dropdown.ts:122]: As noted in the comment for `isMenuGenerator`, there's a quirk with type guarding for `Sentinel` so I shifted the order around.

File: core/field_dropdown.ts
Comment (btw17) [inline core/field_dropdown.ts:125]: Before, `trimOptions` was mutating the deep copied menuGenerator and, when no `ImageProperties` labels were provided, setting a new value for `menuGenerator_`. Assuming there's not an edge case I'm missing where the mutation is needed, this update shifts setting the value for `menuGenerator_` to one point, and it should eliminate the need for a deep copy.

File: core/field_dropdown.ts
Comment (btw17) [inline core/field_dropdown.ts:226]: `this.arrow_` apparently changed at some point since being given `AnyDuringMigration` since it worked when I just removed `AnyDuringMigration`.

**Comment by btw17** [comment]
Looks like there were a few commits from the `master` branch I pulled over into this PR that weren't in `develop`. I'm assuming they'll be merged into `develop` in a bit, at which point I'll clean up this PR.

**Comment by maribethb** [comment]
> Looks like there were a few commits from the `master` branch I pulled over into this PR that weren't in `develop`. I'm assuming they'll be merged into `develop` in a bit, at which point I'll clean up this PR.

Unfortunately this won't be a timely assumption, likely until our next release. It would be best to rebase onto develop.

**Comment by btw17** [comment]
> > Looks like there were a few commits from the `master` branch I pulled over into this PR that weren't in `develop`. I'm assuming they'll be merged into `develop` in a bit, at which point I'll clean up this PR.
> 
> Unfortunately this won't be a timely assumption, likely until our next release. It would be best to rebase onto develop.

If `develop` won't be until the next major release, would it be fine to merge this into `master` instead of `develop` (once commits are cleaned up accordingly)? The idea is there wouldn't be a noticeable functional change here.

**Comment by maribethb** [comment]
> would it be fine to merge this into `master` instead of `develop` (once commits are cleaned up accordingly)? 

Unfortunately no, that's just not how our workflow works. All PRs should go into develop as that is what we test with before merging a rollup of changes into master. Generally, master is equivalent with whatever code is currently on npm. Pushing something straight to master would break that assumption.

> If `develop` won't be until the next major release, 

`master` won't be merged into `develop` until the next release (major, minor, or patch*). That is another reason we can't push to master, because everyone else is working from develop. Someone else could change this line in develop, creating a merge conflict at the time of our next release.


* We don't always merge master back into develop when doing a cherry-pick patch release, but my point is that it's not just during major releases.

File: core/field_dropdown.ts
Comment (btw17) [reply core/field_dropdown.ts:117]: For function overloading in TS, the final function's parameters need to support all of the previous functions' parameters. The final constructor supports `MenuGenerator|Sentinel` since those are the cases that need to be handled in code within the constructor. It's also ultimately the limit of what can be provided, so a user *could* provide both a `menuGenerator: Sentinel` and a validator.

The first two constructors here just guide what combinations should be provided. Without providing any parameters, it'll suggest providing a `MenuGenerator`. If a `Sentinel` is provided, it'll suggest not providing a validator or config. Since Sentinel is only for subclasses, it seems like most users won't want to provide it.

File: core/field_dropdown.ts
Comment (btw17) [reply core/field_dropdown.ts:689]: Will do! I think originally I had the return on line 714, though the formatter probably bumped it to a new line.

File: core/field_dropdown.ts
Comment (btw17) [reply core/field_dropdown.ts:368]: Before, an empty array or a generator that produces an empty array could also be provided. Good point, though - if we're assuming the array should be non-empty then we should throw an error if it isn't.

File: core/field_dropdown.ts
Comment (btw17) [reply core/field_dropdown.ts:127]: Originally, I tried keeping the resulting values identical to the previous implementation in mind of a minor release. Since this will probably be included in a major release, though, I think it might be worth making `prefixField` and `suffixField` accept undefined instead of null. `null` is in a weird spot where it's mostly symbolic now, perhaps conveying an explicit choice to provide a non-value rather than choosing to not provide a value.

I'll make the update in favor of undefined since it keeps the code more concise, though I can adjust either way.

Comment (btw17) [reply core/field_dropdown.ts:127]: Actually, updating `prefixField` and `suffixField` would require updating the `Field` parent class and impacted files so that'd be a pretty expansive change. Would it be fine to leave this as is for now? Ultimately I think those values probably should be just optional strings rather than string|null, though that's probably for a future PR.

File: core/field_dropdown.ts
Comment (btw17) [reply core/field_dropdown.ts:634]: Done!

File: core/field_dropdown.ts
Comment (btw17) [reply core/field_dropdown.ts:701]: Fixed!

File: core/field_dropdown.ts
Comment (btw17) [reply core/field_dropdown.ts]: Done!

File: core/field_dropdown.ts
Comment (btw17) [reply core/field_dropdown.ts]: Good point - updated!

**Comment by btw17** [comment]
I made most of the requested updates and cleaned up the branch to no longer include those commits from `master`. The only thing left outstanding in the discussion of `null` vs `undefined` and if there was any further thoughts about overloading the constructor.

File: core/field_dropdown.ts
Comment (btw17) [inline core/field_dropdown.ts]: Oh! I just noticed that `validateOptions` checks for a non-empty array. I was incorrect before in my assumption that the generator function and array could be empty. Updated again!

File: core/field_dropdown.ts
Comment (btw17) [reply core/field_dropdown.ts:368]: I just noticed that `validateOptions` checks for a non-empty array. I was incorrect before in my assumption that the generator function and array could be empty. Updated again!

File: core/field_dropdown.ts
Comment (BeksOmega) [inline core/field_dropdown.ts:49]: Pretty sure this isn't actually a problem anymore (I can view this file just fine on my android phone). Could you have this just use the small carrot string + remove the comment?

File: core/field_dropdown.ts
Comment (BeksOmega) [inline core/field_dropdown.ts:116]: nit: can you put each param on a separate line (here and below)? Related to: https://google.github.io/styleguide/jsguide.html#formatting-where-to-break and https://github.com/google/google-java-format/wiki/The-Rectangle-Rule

File: core/field_dropdown.ts
Comment (BeksOmega) [reply core/field_dropdown.ts:698]: Could you still move the comment so that which part is the predicate is a bit easier to parse? I think either put it right above the if-statement, or right about the return statement.

File: core/field_dropdown.ts
Comment (BeksOmega) [inline core/field_dropdown.ts:669]: I reeeeally like this change :D

File: core/field_dropdown.ts
Comment (BeksOmega) [inline core/field_dropdown.ts:720]: Yay pure programming instead of mutation!

File: core/field_dropdown.ts
Comment (BeksOmega) [inline core/field_dropdown.ts:394]: Yes, love this change!

File: core/field_variable.ts
Comment (BeksOmega) [inline core/field_variable.ts:91]: Mmm this is kind of unfortunate (having to cast with `as`), but I think this is the right thing to do here, b/c sadly expecting a `FieldVariable` means it isn't actually a subtype :/

File: core/field_dropdown.ts
Comment (BeksOmega) [reply core/field_dropdown.ts:119]: I think this combination of constructors makes sense!

**Review by BeksOmega — APPROVED** [review]
LGTM! I wrote down a few nits, but they're all optional in my mind, so I'm going ahead and approving this =) Will merge Tuesday after making sure Maribeth is cool with it.

File: core/field_dropdown.ts
Comment (btw17) [reply core/field_dropdown.ts]: Done!

File: core/field_dropdown.ts
Comment (btw17) [reply core/field_dropdown.ts]: Done!

File: core/field_dropdown.ts
Comment (btw17) [reply core/field_dropdown.ts:670]: I moved the comment to right above the return statement.

Glad you like the change! ⭐

File: core/field_dropdown.ts
Comment (btw17) [reply core/field_dropdown.ts:395]: Condensing code's incredibly satisfying. Glad you like the change! ⭐

File: core/field_variable.ts
Comment (btw17) [reply core/field_variable.ts:91]: Yeah, that was my thought as well. I explored a few different options, like updating `dropdownCreate` to accept generics, though in the end this made the most sense.

Ideally this could be refactored such that `this` isn't a parameter for the base function, though my goal was to avoid impacting the usage of each function.

**Comment by btw17** [comment]
Pulled in this week's `develop` changes and rebased - should be good to go now!

**Review by maribethb — APPROVED** [review]
Thanks for making the changes and thanks to Beka for the additional review!

**Pull request merged**
